### ISSUE TITLE
Multiple change streams cause native crashes on occasion

### ISSUE THREAD

**Issue by whatthehecker** [cbl-dart/cbl-dart#332]
**Describe the bug**

Our Flutter app uses several change listener streams to listen to entities of a certain type on each screen. We do not cache these streams, so our app as a whole may hold dozens of change listener streams, some of which listen to the same thing.
On some occasions when modifying database contents by writing or deleting documents, the app crashes with a native stack trace. We have 3 different possible stacks we receive this way, I will only attach the most recent one for this issue.

We are unsure whether this is a problem in cbl or the C API.

**To Reproduce**

Only happens occasionally and we are unsure whether there is a general setup to reproduce.
For now we suspect that: 1. multiple change streams must be active on the underlying database and 2. a database change has to occur.

**Expected behavior**

Multiple change streams can be open at the same time without any crashes occurring.

**Screenshots**
<details>
<summary>Symbolized stack trace</summary>

```
********** Crash dump: **********
Build fingerprint: 'google/blueline/blueline:12/SP1A.210812.016.C1/8029091:user/release-keys'
#00 0x000000000005df08 /data/app/~~HCS34imsYEbu6uQVRN9bmA==/some.app-T9fVAJyNFxNZdtVlRU82Fg==/lib/arm64/libcblite.so (BuildId: 7ddf6f30ffcaa330988cc2e245182048c8c0c2ad)
                                                                                                                                                           int std::__ndk1::__cxx_atomic_fetch_add<int>(std::__ndk1::__cxx_atomic_base_impl<int>*, int, std::__ndk1::memory_order)
                                                                                                                                                           /home/couchbase/jenkins/tools/android-sdk/ndk/21.4.7075529/toolchains/llvm/prebuilt/linux-x86_64/sysroot/usr/include/c++/v1/atomic:1014:12
#01 0x000000000005dda8 /data/app/~~HCS34imsYEbu6uQVRN9bmA==/some.app-T9fVAJyNFxNZdtVlRU82Fg==/lib/arm64/libcblite.so (BuildId: 7ddf6f30ffcaa330988cc2e245182048c8c0c2ad)
                                                                                                                                                           std::__ndk1::__atomic_base<int, true>::fetch_add(int, std::__ndk1::memory_order)
                                                                                                                                                           /home/couchbase/jenkins/tools/android-sdk/ndk/21.4.7075529/toolchains/llvm/prebuilt/linux-x86_64/sysroot/usr/include/c++/v1/atomic:1575:17
#02 0x000000000005dd70 /data/app/~~HCS34imsYEbu6uQVRN9bmA==/some.app-T9fVAJyNFxNZdtVlRU82Fg==/lib/arm64/libcblite.so (BuildId: 7ddf6f30ffcaa330988cc2e245182048c8c0c2ad)
                                                                                                                                                           std::__ndk1::__atomic_base<int, true>::operator++()
                                                                                                                                                           /home/couchbase/jenkins/tools/android-sdk/ndk/21.4.7075529/toolchains/llvm/prebuilt/linux-x86_64/sysroot/usr/include/c++/v1/atomic:1612:57
#03 0x0000000000448e0c /data/app/~~HCS34imsYEbu6uQVRN9bmA==/some.app-T9fVAJyNFxNZdtVlRU82Fg==/lib/arm64/libcblite.so (BuildId: 7ddf6f30ffcaa330988cc2e245182048c8c0c2ad)
                                                                                                                                                           fleece::RefCounted::_retain() const
                                                                                                                                                           /home/couchbase/jenkins/workspace/couchbase-lite-c-android/build_android_arm64/../couchbase-lite-c/vendor/couchbase-lite-core/Networking/BLIP/../../vendor/fleece/Fleece/Support/RefCounted.hh:49:57
                                                                                                                                                           litecore::actor::Actor* fleece::retain<litecore::actor::Actor>(litecore::actor::Actor*)
                                                                                                                                                           /home/couchbase/jenkins/workspace/couchbase-lite-c-android/build_android_arm64/../couchbase-lite-c/vendor/couchbase-lite-core/Networking/BLIP/../../vendor/fleece/Fleece/Support/RefCounted.hh:72:0
                                                                                                                                                           litecore::actor::ThreadedMailbox::enqueue(char const*, std::__ndk1::function<void ()> const&)
                                                                                                                                                           /home/couchbase/jenkins/workspace/couchbase-lite-c-android/build_android_arm64/../couchbase-lite-c/vendor/couchbase-lite-core/LiteCore/Support/ThreadedMailbox.cc:142:0
#04 0x000000000016f620 /data/app/~~HCS34imsYEbu6uQVRN9bmA==/some.app-T9fVAJyNFxNZdtVlRU82Fg==/lib/arm64/libcblite.so (BuildId: 7ddf6f30ffcaa330988cc2e245182048c8c0c2ad)
                                                                                                                                                           void litecore::actor::Actor::enqueue<litecore::LiveQuerier>(char const*, void (litecore::LiveQuerier::*)())
                                                                                                                                                           /home/couchbase/jenkins/workspace/couchbase-lite-c-android/build_android_arm64/../couchbase-lite-c/vendor/couchbase-lite-core/LiteCore/Support/Actor.hh:111:22
#05 0x000000000016f4b4 /data/app/~~HCS34imsYEbu6uQVRN9bmA==/some.app-T9fVAJyNFxNZdtVlRU82Fg==/lib/arm64/libcblite.so (BuildId: 7ddf6f30ffcaa330988cc2e245182048c8c0c2ad)
                                                                                                                                                           litecore::LiveQuerier::stop()
                                                                                                                                                           /home/couchbase/jenkins/workspace/couchbase-lite-c-android/build_android_arm64/../couchbase-lite-c/vendor/couchbase-lite-core/LiteCore/Database/LiveQuerier.cc:98:9
#06 0x0000000000125e6c /data/app/~~HCS34imsYEbu6uQVRN9bmA==/some.app-T9fVAJyNFxNZdtVlRU82Fg==/lib/arm64/libcblite.so (BuildId: 7ddf6f30ffcaa330988cc2e245182048c8c0c2ad)
                                                                                                                                                           C4Query::enableObserver(litecore::C4QueryObserverImpl*, bool)
                                                                                                                                                           /home/couchbase/jenkins/workspace/couchbase-lite-c-android/build_android_arm64/../couchbase-lite-c/vendor/couchbase-lite-core/C/c4Query.cc:275:25
#07 0x000000000012778c /data/app/~~HCS34imsYEbu6uQVRN9bmA==/some.app-T9fVAJyNFxNZdtVlRU82Fg==/lib/arm64/libcblite.so (BuildId: 7ddf6f30ffcaa330988cc2e245182048c8c0c2ad)
                                                                                                                                                           litecore::C4QueryObserverImpl::~C4QueryObserverImpl()
                                                                                                                                                           /home/couchbase/jenkins/workspace/couchbase-lite-c-android/build_android_arm64/../couchbase-lite-c/vendor/couchbase-lite-core/C/c4QueryImpl.hh:129:25
#08 0x0000000000127844 /data/app/~~HCS34imsYEbu6uQVRN9bmA==/some.app-T9fVAJyNFxNZdtVlRU82Fg==/lib/arm64/libcblite.so (BuildId: 7ddf6f30ffcaa330988cc2e245182048c8c0c2ad)
                                                                                                                                                           litecore::C4QueryObserverImpl::~C4QueryObserverImpl()
                                                                                                                                                           /home/couchbase/jenkins/workspace/couchbase-lite-c-android/build_android_arm64/../couchbase-lite-c/vendor/couchbase-lite-core/C/c4QueryImpl.hh:127:32
#09 0x0000000000097550 /data/app/~~HCS34imsYEbu6uQVRN9bmA==/some.app-T9fVAJyNFxNZdtVlRU82Fg==/lib/arm64/libcblite.so (BuildId: 7ddf6f30ffcaa330988cc2e245182048c8c0c2ad)
                                                                                                                                                           std::__ndk1::default_delete<C4QueryObserver>::operator()(C4QueryObserver*) const
                                                                                                                                                           /home/couchbase/jenkins/tools/android-sdk/ndk/21.4.7075529/toolchains/llvm/prebuilt/linux-x86_64/sysroot/usr/include/c++/v1/memory:2338:5
#10 0x0000000000097434 /data/app/~~HCS34imsYEbu6uQVRN9bmA==/some.app-T9fVAJyNFxNZdtVlRU82Fg==/lib/arm64/libcblite.so (BuildId: 7ddf6f30ffcaa330988cc2e245182048c8c0c2ad)
                                                                                                                                                           std::__ndk1::unique_ptr<C4QueryObserver, std::__ndk1::default_delete<C4QueryObserver> >::reset(C4QueryObserver*)
                                                                                                                                                           /home/couchbase/jenkins/tools/android-sdk/ndk/21.4.7075529/toolchains/llvm/prebuilt/linux-x86_64/sysroot/usr/include/c++/v1/memory:2593:7
#11 0x00000000000954b4 /data/app/~~HCS34imsYEbu6uQVRN9bmA==/some.app-T9fVAJyNFxNZdtVlRU82Fg==/lib/arm64/libcblite.so (BuildId: 7ddf6f30ffcaa330988cc2e245182048c8c0c2ad)
                                                                                                                                                           std::__ndk1::unique_ptr<C4QueryObserver, std::__ndk1::default_delete<C4QueryObserver> >::~unique_ptr()
                                                                                                                                                           /home/couchbase/jenkins/tools/android-sdk/ndk/21.4.7075529/toolchains/llvm/prebuilt/linux-x86_64/sysroot/usr/include/c++/v1/memory:2547:19
#12 0x000000000009552c /data/app/~~HCS34imsYEbu6uQVRN9bmA==/some.app-T9fVAJyNFxNZdtVlRU82Fg==/lib/arm64/libcblite.so (BuildId: 7ddf6f30ffcaa330988cc2e245182048c8c0c2ad)
                                                                                                                                                           cbl_internal::ListenerToken<void (*)(void*, CBLQuery*, CBLListenerToken*)>::~ListenerToken()
                                                                                                                                                           /home/couchbase/jenkins/workspace/couchbase-lite-c-android/build_android_arm64/../couchbase-lite-c/src/CBLQuery_Internal.hh:198:9
#13 0x00000000000955c8 /data/app/~~HCS34imsYEbu6uQVRN9bmA==/some.app-T9fVAJyNFxNZdtVlRU82Fg==/lib/arm64/libcblite.so (BuildId: 7ddf6f30ffcaa330988cc2e245182048c8c0c2ad)
                                                                                                                                                           cbl_internal::ListenerToken<void (*)(void*, CBLQuery*, CBLListenerToken*)>::~ListenerToken()
                                                                                                                                                           /home/couchbase/jenkins/workspace/couchbase-lite-c-android/build_android_arm64/../couchbase-lite-c/src/CBLQuery_Internal.hh:187:26
#14 0x00000000003d49f8 /data/app/~~HCS34imsYEbu6uQVRN9bmA==/some.app-T9fVAJyNFxNZdtVlRU82Fg==/lib/arm64/libcblite.so (BuildId: 7ddf6f30ffcaa330988cc2e245182048c8c0c2ad)
                                                                                                                                                           fleece::RefCounted::_release() const
                                                                                                                                                           /home/couchbase/jenkins/workspace/couchbase-lite-c-android/build_android_arm64/../couchbase-lite-c/vendor/couchbase-lite-core/vendor/fleece/Fleece/Support/RefCounted.cc:28:13
#15 0x00000000003d4a30 /data/app/~~HCS34imsYEbu6uQVRN9bmA==/some.app-T9fVAJyNFxNZdtVlRU82Fg==/lib/arm64/libcblite.so (BuildId: 7ddf6f30ffcaa330988cc2e245182048c8c0c2ad)
                                                                                                                                                           fleece::release(fleece::RefCounted const*)
                                                                                                                                                           /home/couchbase/jenkins/workspace/couchbase-lite-c-android/build_android_arm64/../couchbase-lite-c/vendor/couchbase-lite-core/vendor/fleece/Fleece/Support/RefCounted.cc:34:19
#16 0x000000000005dc74 /data/app/~~HCS34imsYEbu6uQVRN9bmA==/some.app-T9fVAJyNFxNZdtVlRU82Fg==/lib/arm64/libcblite.so (CBLListener_Remove+52) (BuildId: 7ddf6f30ffcaa330988cc2e245182048c8c0c2ad)
                                                                                                                                                           CBLListener_Remove
                                                                                                                                                           /home/couchbase/jenkins/workspace/couchbase-lite-c-android/build_android_arm64/../couchbase-lite-c/src/CBLBase_CAPI.cc:76:9
#17 0x000000000000db30 /data/app/~~HCS34imsYEbu6uQVRN9bmA==/some.app-T9fVAJyNFxNZdtVlRU82Fg==/lib/arm64/libcblitedart.so (BuildId: ca83ba30c4ae152204433a3065e30d2303ae1eae)
#18 0x000000000000eb00 /data/app/~~HCS34imsYEbu6uQVRN9bmA==/some.app-T9fVAJyNFxNZdtVlRU82Fg==/lib/arm64/libcblitedart.so (CBLDart_AsyncCallback_Close+16) (BuildId: ca83ba30c4ae152204433a3065e30d2303ae1eae)
#19 0x0000000000006db4 [anon:dart-codespace]
```
</details>

**Environment (please complete the following information):**

 - OS: Android
 - `cbl`: 1.0.3
 - `cbl_flutter`: 1.0.0+2
 - `cbl_flutter_ce`: 1.1.0

**Additional context**

As mentioned previously, this is only one of the three different stacks we receive when crashes like these occur. This crash resulted in SIGBUS while the other ones result in SIGSEGV and SIGABRT. If required, I can also add the stack traces for the other two crashes (although this may take a while to reproduce them on the newest version).

Seeing as these crashes happen seemingly at random, I suspect there is a data race happening somewhere in the C API which may or may not be caused by the way the Dart code interacts with it.

**Comment by blaugold** [comment]
From the stack trace, it looks like the crash is triggered when the stream is canceled (`CBLListener_Remove ` is called from the Dart side when shutting a stream down).

I found a bug which could result in `CBLListener_Remove` being additionally called by the stream's native finalizer.
#333 is a fix for that. Since this is hard to replicate, I'm not completely sure that it is going to fix this issue, but reading through the CBL C code I couldn't find a different explanation either.

**Comment by blaugold** [comment]
I'll publish a new release, with the fix mentioned above, later today. If this issue is not resolved by it, please reopen it.

**Comment by blaugold** [comment]
`cbl` 1.0.4 has been published.

**Comment by roberthecker** [comment]
Thanks for the quick response and fix. Sadly I now get a different crash than before:
<details>
<summary>Symbolized stack trace</summary>

```
********** Crash dump: **********
Build fingerprint: 'google/blueline/blueline:12/SP1A.210812.016.C1/8029091:user/release-keys'
#00 0x00000000000ae100 /data/app/~~xt4Xfdz8CllJLGL7QHyuiQ==/some.app-yB1-UseIUZWGgoCIqjwe5A==/lib/arm64/libcblite.so (BuildId: 7ddf6f30ffcaa330988cc2e245182048c8c0c2ad)
                                                                                                                                                           litecore::LogLevel std::__ndk1::__cxx_atomic_load<litecore::LogLevel>(std::__ndk1::__cxx_atomic_base_impl<litecore::LogLevel> const*, std::__ndk1::memory_order)
                                                                                                                                                           /home/couchbase/jenkins/tools/android-sdk/ndk/21.4.7075529/toolchains/llvm/prebuilt/linux-x86_64/sysroot/usr/include/c++/v1/atomic:970:12
#01 0x00000000000ae090 /data/app/~~xt4Xfdz8CllJLGL7QHyuiQ==/some.app-yB1-UseIUZWGgoCIqjwe5A==/lib/arm64/libcblite.so (BuildId: 7ddf6f30ffcaa330988cc2e245182048c8c0c2ad)
                                                                                                                                                           std::__ndk1::__atomic_base<litecore::LogLevel, false>::load(std::__ndk1::memory_order) const
                                                                                                                                                           /home/couchbase/jenkins/tools/android-sdk/ndk/21.4.7075529/toolchains/llvm/prebuilt/linux-x86_64/sysroot/usr/include/c++/v1/atomic:1487:17
#02 0x00000000000ae064 /data/app/~~xt4Xfdz8CllJLGL7QHyuiQ==/some.app-yB1-UseIUZWGgoCIqjwe5A==/lib/arm64/libcblite.so (BuildId: 7ddf6f30ffcaa330988cc2e245182048c8c0c2ad)
                                                                                                                                                           std::__ndk1::__atomic_base<litecore::LogLevel, false>::operator litecore::LogLevel() const
                                                                                                                                                           /home/couchbase/jenkins/tools/android-sdk/ndk/21.4.7075529/toolchains/llvm/prebuilt/linux-x86_64/sysroot/usr/include/c++/v1/atomic:1491:53
#03 0x00000000000ae018 /data/app/~~xt4Xfdz8CllJLGL7QHyuiQ==/some.app-yB1-UseIUZWGgoCIqjwe5A==/lib/arm64/libcblite.so (BuildId: 7ddf6f30ffcaa330988cc2e245182048c8c0c2ad)
                                                                                                                                                           litecore::LogDomain::willLog(litecore::LogLevel) const
                                                                                                                                                           /home/couchbase/jenkins/workspace/couchbase-lite-c-android/build_android_arm64/../couchbase-lite-c/vendor/couchbase-lite-core/LiteCore/Support/Logging.hh:91:61
#04 0x00000000000a9828 /data/app/~~xt4Xfdz8CllJLGL7QHyuiQ==/some.app-yB1-UseIUZWGgoCIqjwe5A==/lib/arm64/libcblite.so (BuildId: 7ddf6f30ffcaa330988cc2e245182048c8c0c2ad)
                                                                                                                                                           litecore::Logging::willLog(litecore::LogLevel) const
                                                                                                                                                           /home/couchbase/jenkins/workspace/couchbase-lite-c-android/build_android_arm64/../couchbase-lite-c/vendor/couchbase-lite-core/LiteCore/Support/Logging.hh:234:84
#05 0x000000000016f468 /data/app/~~xt4Xfdz8CllJLGL7QHyuiQ==/some.app-yB1-UseIUZWGgoCIqjwe5A==/lib/arm64/libcblite.so (BuildId: 7ddf6f30ffcaa330988cc2e245182048c8c0c2ad)
                                                                                                                                                           litecore::LiveQuerier::stop()
                                                                                                                                                           /home/couchbase/jenkins/workspace/couchbase-lite-c-android/build_android_arm64/../couchbase-lite-c/vendor/couchbase-lite-core/LiteCore/Database/LiveQuerier.cc:94:13
#06 0x0000000000125e6c /data/app/~~xt4Xfdz8CllJLGL7QHyuiQ==/some.app-yB1-UseIUZWGgoCIqjwe5A==/lib/arm64/libcblite.so (BuildId: 7ddf6f30ffcaa330988cc2e245182048c8c0c2ad)
                                                                                                                                                           C4Query::enableObserver(litecore::C4QueryObserverImpl*, bool)
                                                                                                                                                           /home/couchbase/jenkins/workspace/couchbase-lite-c-android/build_android_arm64/../couchbase-lite-c/vendor/couchbase-lite-core/C/c4Query.cc:275:25
#07 0x000000000012778c /data/app/~~xt4Xfdz8CllJLGL7QHyuiQ==/some.app-yB1-UseIUZWGgoCIqjwe5A==/lib/arm64/libcblite.so (BuildId: 7ddf6f30ffcaa330988cc2e245182048c8c0c2ad)
                                                                                                                                                           litecore::C4QueryObserverImpl::~C4QueryObserverImpl()
                                                                                                                                                           /home/couchbase/jenkins/workspace/couchbase-lite-c-android/build_android_arm64/../couchbase-lite-c/vendor/couchbase-lite-core/C/c4QueryImpl.hh:129:25
#08 0x0000000000127844 /data/app/~~xt4Xfdz8CllJLGL7QHyuiQ==/some.app-yB1-UseIUZWGgoCIqjwe5A==/lib/arm64/libcblite.so (BuildId: 7ddf6f30ffcaa330988cc2e245182048c8c0c2ad)
                                                                                                                                                           litecore::C4QueryObserverImpl::~C4QueryObserverImpl()
                                                                                                                                                           /home/couchbase/jenkins/workspace/couchbase-lite-c-android/build_android_arm64/../couchbase-lite-c/vendor/couchbase-lite-core/C/c4QueryImpl.hh:127:32
#09 0x0000000000097550 /data/app/~~xt4Xfdz8CllJLGL7QHyuiQ==/some.app-yB1-UseIUZWGgoCIqjwe5A==/lib/arm64/libcblite.so (BuildId: 7ddf6f30ffcaa330988cc2e245182048c8c0c2ad)
                                                                                                                                                           std::__ndk1::default_delete<C4QueryObserver>::operator()(C4QueryObserver*) const
                                                                                                                                                           /home/couchbase/jenkins/tools/android-sdk/ndk/21.4.7075529/toolchains/llvm/prebuilt/linux-x86_64/sysroot/usr/include/c++/v1/memory:2338:5
#10 0x0000000000097434 /data/app/~~xt4Xfdz8CllJLGL7QHyuiQ==/some.app-yB1-UseIUZWGgoCIqjwe5A==/lib/arm64/libcblite.so (BuildId: 7ddf6f30ffcaa330988cc2e245182048c8c0c2ad)
                                                                                                                                                           std::__ndk1::unique_ptr<C4QueryObserver, std::__ndk1::default_delete<C4QueryObserver> >::reset(C4QueryObserver*)
                                                                                                                                                           /home/couchbase/jenkins/tools/android-sdk/ndk/21.4.7075529/toolchains/llvm/prebuilt/linux-x86_64/sysroot/usr/include/c++/v1/memory:2593:7
#11 0x00000000000954b4 /data/app/~~xt4Xfdz8CllJLGL7QHyuiQ==/some.app-yB1-UseIUZWGgoCIqjwe5A==/lib/arm64/libcblite.so (BuildId: 7ddf6f30ffcaa330988cc2e245182048c8c0c2ad)
                                                                                                                                                           std::__ndk1::unique_ptr<C4QueryObserver, std::__ndk1::default_delete<C4QueryObserver> >::~unique_ptr()
                                                                                                                                                           /home/couchbase/jenkins/tools/android-sdk/ndk/21.4.7075529/toolchains/llvm/prebuilt/linux-x86_64/sysroot/usr/include/c++/v1/memory:2547:19
#12 0x000000000009552c /data/app/~~xt4Xfdz8CllJLGL7QHyuiQ==/some.app-yB1-UseIUZWGgoCIqjwe5A==/lib/arm64/libcblite.so (BuildId: 7ddf6f30ffcaa330988cc2e245182048c8c0c2ad)
                                                                                                                                                           cbl_internal::ListenerToken<void (*)(void*, CBLQuery*, CBLListenerToken*)>::~ListenerToken()
                                                                                                                                                           /home/couchbase/jenkins/workspace/couchbase-lite-c-android/build_android_arm64/../couchbase-lite-c/src/CBLQuery_Internal.hh:198:9
#13 0x00000000000955c8 /data/app/~~xt4Xfdz8CllJLGL7QHyuiQ==/some.app-yB1-UseIUZWGgoCIqjwe5A==/lib/arm64/libcblite.so (BuildId: 7ddf6f30ffcaa330988cc2e245182048c8c0c2ad)
                                                                                                                                                           cbl_internal::ListenerToken<void (*)(void*, CBLQuery*, CBLListenerToken*)>::~ListenerToken()
                                                                                                                                                           /home/couchbase/jenkins/workspace/couchbase-lite-c-android/build_android_arm64/../couchbase-lite-c/src/CBLQuery_Internal.hh:187:26
#14 0x00000000003d49f8 /data/app/~~xt4Xfdz8CllJLGL7QHyuiQ==/some.app-yB1-UseIUZWGgoCIqjwe5A==/lib/arm64/libcblite.so (BuildId: 7ddf6f30ffcaa330988cc2e245182048c8c0c2ad)
                                                                                                                                                           fleece::RefCounted::_release() const
                                                                                                                                                           /home/couchbase/jenkins/workspace/couchbase-lite-c-android/build_android_arm64/../couchbase-lite-c/vendor/couchbase-lite-core/vendor/fleece/Fleece/Support/RefCounted.cc:28:13
#15 0x00000000003d4a30 /data/app/~~xt4Xfdz8CllJLGL7QHyuiQ==/some.app-yB1-UseIUZWGgoCIqjwe5A==/lib/arm64/libcblite.so (BuildId: 7ddf6f30ffcaa330988cc2e245182048c8c0c2ad)
                                                                                                                                                           fleece::release(fleece::RefCounted const*)
                                                                                                                                                           /home/couchbase/jenkins/workspace/couchbase-lite-c-android/build_android_arm64/../couchbase-lite-c/vendor/couchbase-lite-core/vendor/fleece/Fleece/Support/RefCounted.cc:34:19
#16 0x000000000005dc74 /data/app/~~xt4Xfdz8CllJLGL7QHyuiQ==/some.app-yB1-UseIUZWGgoCIqjwe5A==/lib/arm64/libcblite.so (CBLListener_Remove+52) (BuildId: 7ddf6f30ffcaa330988cc2e245182048c8c0c2ad)
                                                                                                                                                           CBLListener_Remove
                                                                                                                                                           /home/couchbase/jenkins/workspace/couchbase-lite-c-android/build_android_arm64/../couchbase-lite-c/src/CBLBase_CAPI.cc:76:9
#17 0x000000000000db30 /data/app/~~xt4Xfdz8CllJLGL7QHyuiQ==/some.app-yB1-UseIUZWGgoCIqjwe5A==/lib/arm64/libcblitedart.so (BuildId: ca83ba30c4ae152204433a3065e30d2303ae1eae)
#18 0x000000000000eb00 /data/app/~~xt4Xfdz8CllJLGL7QHyuiQ==/some.app-yB1-UseIUZWGgoCIqjwe5A==/lib/arm64/libcblitedart.so (CBLDart_AsyncCallback_Close+16) (BuildId: ca83ba30c4ae152204433a3065e30d2303ae1eae)
#19 0x0000000000006db4 [anon:dart-codespace]
```

</details>

This is one of the other possible stack traces I mentioned in the original report. I'll look into whether the original crash does not occur anymore in our configuration.

**Comment by roberthecker** [comment]
I also cannot reopen this issue myself; if needed I can create a new one.

**Comment by blaugold** [comment]
I'm sorry you're still having to deal with this issue, and I reopened it.

Both stack traces point to a double free issue, so I suspect the root cause is still the same.

If you can capture file logs, those are often very helpful. Let me know if you need help with that.

**Comment by roberthecker** [comment]
I am unsure what kind of logs you need, so a pointer in the right direction would be very appreciated.

**Comment by blaugold** [comment]
The [`FileLogger`](https://pub.dev/documentation/cbl/latest/cbl/FileLogger-class.html) can be configured to record log messages straight to disk. This makes it easy to capture even fine-grained log messages and has less overhead than going through Dart.

```dart
Database.log.file
  ..config = LogFileConfiguration(
    directory: ...,
    // It's easier to use plain text in a dev environment because no decoding from the binary format is necessary.
    // The binary format is smaller and faster to write.
    usePlainText: true,
  )
  ..level = LogLevel.verbose;
```

The log directory will contain one file for each log level.

**Comment by blaugold** [comment]
Never mind about those logs. I'm pretty sure I found the bug in CBL C, and I'll file a bug report there.

**Comment by blaugold** [comment]
This should be fixed in the 2.1.0 release of `cbl_dart`, `cbl_flutter_ce` and `cbl_flutter_ee`. Please comment if the issue persists.
